### PR TITLE
Switch to npm

### PR DESCRIPTION
**Pull request by callumlocke**
## ~~On hold~~

Problem: npm5, when running on CircleCI, seems to hit lots of errors in the form `EINVAL: invalid argument, chown ...`

This seems to be a known issue in the [CircleCI forums](https://discuss.circleci.com/t/getting-a-weird-einval-error-with-npm-on-circleci-2-0/13465). Possibly caused by [this npm bug](https://github.com/npm/npm/issues/16896).

I've seen one workaround involving making a custom Docker image that creates a new user, so npm doesn't have to install things under `/root`. But I think it's better to stick with the standard `node:8` image if possible.

So we should probably just wait for a fix to the [npm bug](https://github.com/npm/npm/issues/16896).

---

## Done

In this PR:

- update dependencies
- switch from yarn to npm
- specify node 8 and npm 5 in package.json (this is just advisory; it has no real effect except in certain situations like Heroku)
- use circle's own npm image (thanks aendrew)
- use a hash of package-lock.json as the key for caching node_modules between builds
- get rid of the `--allow-root` from bower, since it doesn't run as sudo on this new docker image

### CODE REVIEW

**Comment by aendrew** [comment]
@callumlocke According to [this](https://github.com/npm/npm/issues/16896#issuecomment-312930312) it'll be fixed in today's 5.1.0 release. :tada:

**Comment by callumlocke** [comment]
ah this is interesting... it **only** builds without cache. if you try a regular build, it fails when it tries to restore the bower cache, saying permissions are denied

**Comment by callumlocke** [comment]
ok this finally works.

the last two builds show a non-cache build and a from-cache build respectively (and the entire "install deps" step is only 6 seconds on the from-cache build, which is good)

after this PR I want to try removing or refinding the `npm rebuild` step so we can shave off 15 seconds. but I'll try that separately.

File: package.json [hunk 1]
```diff
@@ -61,27 +61,27 @@
...
+    "run-sequence": "^2.0.0",
     "serve-static": "^1.12.3",
     "vinyl-source-stream": "^1.1.0",
     "watchify": "^3.9.0"
   },
   "engines": {
-    "node": ">=7.5",
-    "yarn": ">=0.23.2"
+    "node": ">=8",
+    "npm": ">=5"
```
Comment (aendrew) [inline package.json]: Might be worth explicitly specifying `>=5.1.x` so we get an explicit warning if one of the broken versions is being used.

Comment (callumlocke) [inline package.json]: where would we get a warning?

**Pull request merged**